### PR TITLE
Improve stack slots object references handling.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: runtime/src/main/cpp/Memory.cpp [hunk 1]
```diff
@@ -2359,12 +2288,8 @@ OBJ_GETTER(ReadHeapRefLocked, ObjHeader** location, int32_t* spinlock) {
   MEMORY_LOG("ReadHeapRefLocked: %p\n", location)
   lock(spinlock);
   ObjHeader* value = *location;
-  // We do not use UpdateRef() here to avoid having ReleaseRef() on return slot under the lock.
-  if (value != nullptr) {
-    AddStackRef(value);
-  }
   unlock(spinlock);
-  updateReturnRefAdded(OBJ_RESULT, value);
+  UpdateReturnRef(OBJ_RESULT, value);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]:
```kotlin
import kotlin.native.concurrent.*
import platform.posix.*

@SharedImmutable
val ref = AtomicReference<Any?>(Any().freeze())
    
fun main() {
    val worker = Worker.start()
    val future = worker.execute(TransferMode.SAFE, { null }) {
        val value = ref.value
        sleep(2)
        println(value)
    }
    sleep(1)
    ref.value = null
    kotlin.native.internal.GC.collect()

    future.consume {}
    worker.requestTermination().consume {}
}
```

File: backend.native/tests/runtime/workers/worker10.kt [hunk 2]
```diff
@@ -78,4 +78,24 @@ val topSharedData = Data(43)
 
     worker.requestTermination().result
     println("OK")
+}
+
+val ref = AtomicReference<Any?>(Any().freeze())
+
+val semaphore = AtomicInt(0)
+
+@Test fun runTest2() {
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/runtime/workers/worker10.kt]: This test doesn't reproduce [the original bug](https://github.com/JetBrains/kotlin-native/pull/2912#discussion_r278066070) on [cb7e81a7](https://github.com/JetBrains/kotlin-native/commit/cb7e81a7e20f9efcd81cf4a94660e488da933324).

File: runtime/src/main/cpp/Memory.cpp [hunk 3]
```diff
@@ -2359,13 +2272,20 @@ OBJ_GETTER(ReadHeapRefLocked, ObjHeader** location, int32_t* spinlock) {
...
   ObjHeader* value = *location;
-  // We do not use UpdateRef() here to avoid having ReleaseRef() on return slot under the lock.
-  if (value != nullptr) {
-    AddStackRef(value);
-  }
   unlock(spinlock);
-  updateReturnRefAdded(OBJ_RESULT, value);
-  return value;
+  if (value != nullptr)
+    rememberNewContainer(value->container());
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: It's too late to do this after `unlock`.
Another thread may have changed `*location` and performed GC by this moment.
`IncrementRC` must be done under lock.

File: runtime/src/main/cpp/Memory.cpp [hunk 4]
```diff
@@ -1554,6 +1489,16 @@ void garbageCollect(MemoryState* state, bool force) {
   GC_LOG("<<< GC: toFree %d toRelease %d\n", state->toFree->size(), state->toRelease->size())
 }
 
+void rememberNewContainer(ContainerHeader* container) {
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: This operation should also be applied to every reference read from the "heap" that could be released by other thread, e.g.

* References from `StableRef`
* References from `KRefSharedHolder`
* Values of `MutableData.buffer`

File: runtime/src/main/cpp/Memory.cpp [hunk 5]
```diff
@@ -1476,6 +1409,8 @@ void processDecrements(MemoryState* state) {
      toRelease->pop_back();
      if (isMarkedAsRemoved(container))
        continue;
+     if (container->shareable())
+       container = realShareableContainer(container);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: Why not enqueue `ObjHeader*` instead?

File: runtime/src/main/cpp/Memory.cpp [hunk 6]
```diff
@@ -1554,6 +1489,16 @@ void garbageCollect(MemoryState* state, bool force) {
   GC_LOG("<<< GC: toFree %d toRelease %d\n", state->toFree->size(), state->toRelease->size())
 }
 
+void rememberNewContainer(ContainerHeader* container) {
+  if (container == nullptr) return;
+  // Instances can be allocated before actual runtime init - be prepared for that.
+  if (memoryState != nullptr) {
+    IncrementRC</* Atomic = */ false>(container);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: > `/* Atomic = */ false`

Doesn't seem correct when reading from `AtomicReference`.

File: backend.native/tests/runtime/workers/worker10.kt (hunk 2 above)
Comment (olonho) [reply backend.native/tests/runtime/workers/worker10.kt]: Hopefully now it does.

File: runtime/src/main/cpp/Memory.cpp (hunk 5 above)
Comment (olonho) [reply runtime/src/main/cpp/Memory.cpp]: Not a big deal, I guess. We generally try to use containers in MM algorithms.

File: runtime/src/main/cpp/Memory.cpp (hunk 6 above)
Comment (olonho) [reply runtime/src/main/cpp/Memory.cpp]: Yeah, noticed that as well.

File: runtime/src/main/cpp/Memory.cpp (hunk 4 above)
Comment (SvyatoslavScherbina) [reply runtime/src/main/cpp/Memory.cpp]: + `Konan_WeakReferenceCounter_get`

Not sure if I found all such references.

File: runtime/src/main/cpp/Memory.cpp [hunk 7]
```diff
@@ -555,6 +556,15 @@ inline void unlock(KInt* spinlock) {
 
 } // namespace
 
+ObjHeader* KRefSharedHolder::ref() const {
+  verifyRefOwner();
+#if USE_GC
+  if (obj_ != nullptr)
+    rememberNewContainer(obj_->container());
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: There are usages of `KRefSharedHolder::ref()` that don't require `rememberNewContainer`.
Most of them now becomes suboptimal, but few of the usages are now incorrect (e.g. `KotlinBase._tryRetain`).

So we might have two variants of this function, e.g. `ref` and `localRef`, the second one is not supposed to outlive the holder and thus could be used when the reference doesn't escape.

Comment (olonho) [reply runtime/src/main/cpp/Memory.cpp]: Seems that this change makes more harm then good. Let's think over if this behavior is indeed important.

File: runtime/src/main/cpp/Memory.cpp [hunk 8]
```diff
@@ -559,9 +559,10 @@ inline void unlock(KInt* spinlock) {
 ObjHeader* KRefSharedHolder::ref() const {
   verifyRefOwner();
 #if USE_GC
-  if (obj_ != nullptr)
+  if (obj_ != nullptr && owner_ != memoryState)
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: This is wrong. Frozen object may be released by other thread.

File: backend.native/tests/runtime/workers/worker10.kt [hunk 9]
```diff
@@ -118,4 +119,28 @@ val semaphore = AtomicInt(0)
...
+    assertEquals(expected, weak.get())
+}
+
+val stableHolder = StableRef.create(("hello" to "world").freeze())
+
+@Test fun runTest4() {
+    val worker = Worker.start()
+    semaphore.value = 0
+    val future = worker.execute(TransferMode.SAFE, { WeakReference(stableHolder.get()) }) {
+        ensureWeakIs(it, "hello" to "world")
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/runtime/workers/worker10.kt]: This test doesn't reproduce the original issue.
If this worker task is modified to be like that in previous tests, then the issue gets reproduced.

File: runtime/src/main/cpp/Memory.cpp (hunk 8 above)
Comment (SvyatoslavScherbina) [reply runtime/src/main/cpp/Memory.cpp]: As we've discussed earlier, this should be reverted.

**Pull request merged**